packages/verified-fetch/src/verified-fetch.ts

```diff
@@ -147,345 +143,154 @@ export class VerifiedFetch {
-        // response header is set so the user knows to process it differently
-        const obj = ipldDagCbor.decode(block)
-        body = ipldDagJson.encode(obj)
-      } catch (err) {
-        this.log.error('could not transform %c to application/vnd.ipld.dag-json', err)
-        return notAcceptableResponse(resource)
-      }
+    if (contentDisposition != null) {
+      response.headers.set('Content-Disposition', contentDisposition)
     } else {
-      try {
-        body = dagCborToSafeJSON(block)
-      } catch (err) {
-        if (accept === 'application/json') {
-          this.log('could not decode DAG-CBOR as JSON-safe, but the client sent "Accept: application/json"', err)
-
-          return notAcceptableResponse(resource)
-        }
-
-        this.log('could not decode DAG-CBOR as JSON-safe, falling back to `application/octet-stream`', err)
-        body = block
-      }
+      this.log.trace('no content disposition specified')
     }
 
-    const response = okResponse(resource, body)
-
-    if (accept == null) {
-      accept = body instanceof Uint8Array ? 'application/octet-stream' : 'application/json'
+    if (cid != null && response.headers.get('etag') == null) {
+      response.headers.set('etag', getETag({ cid, reqFormat, weak: false }))
     }
 
-    response.headers.set('content-type', accept)
-    setIpfsRoots(response, ipfsRoots)
+    if (protocol != null) {
+      setCacheControlHeader({ response, ttl, protocol })
+    }
+    if (ipfsPath != null) {
+      response.headers.set('X-Ipfs-Path', ipfsPath)
+    }
 
     return response
   }
 
-  private async handleDagPb ({ cid, path, resource, session, options, withServerTiming }: FetchHandlerFunctionArg): Promise<Response> {
-    let redirected = false
-    const byteRangeContext = new ByteRangeContext(this.helia.logger, options?.headers)
-    const blockstore = this.getBlockstore(cid, resource, session, options)
-    const pathDetails = await this.handleServerTiming('path-walking', '', async () => handlePathWalking({ cid, path, resource, options, blockstore, log: this.log, withServerTiming }), withServerTiming)
+  /**
+   * Runs plugins in a loop. After each plugin that returns `null` (partial/no final),
+   * we re-check `canHandle()` for all plugins in the next iteration if the context changed.
+   */
+  private async runPluginPipeline (context: PluginContext, maxPasses: number = 3): Promise<Response | undefined> {
+    let finalResponse: Response | undefined
+    let passCount = 0
+    const pluginsUsed = new Set<string>()
+
+    let prevModificationId = context.modified
+
+    while (passCount < maxPasses) {
+      this.log(`Starting pipeline pass #${passCount + 1}`)
+      passCount++
+
+      // gather plugins that say they can handle the *current* context, but haven't been used yet
+      const readyPlugins = this.plugins.filter(p => !pluginsUsed.has(p.constructor.name)).filter(p => p.canHandle(context))
+      if (readyPlugins.length === 0) {
```

Comment:
When none of the plugins pass the `canHandle` check, why do we then fall back to choosing plugins based on code? 

Why not also filter plugins based on code in the [line above](https://github.com/ipfs/helia-verified-fetch/pull/169/files#diff-12157866119b95f1b476d4e54ac9d3f22a24cc7e7af1a85f251c00c4dfbefcc6R230)?

Reply:
> When none of the plugins pass the canHandle check, why do we then fall back to choosing plugins based on code?

there were some cases in which all plugins returned false for canHandle, but gateway conformance expected some failure state.. I need to dive in deeper. Ideally there would be no `codes` fallback. I think if we break up the plugins more this will naturally resolve itself.

> Why not also filter plugins based on code in the [line above](https://github.com/ipfs/helia-verified-fetch/pull/169/files#diff-12157866119b95f1b476d4e54ac9d3f22a24cc7e7af1a85f251c00c4dfbefcc6R230)?

That link isn't loading for me so i'm not sure

Reply:
I think you're asking: why not also filter plugins that have a particular codec code set that they can handle.. we absolutely could